### PR TITLE
ESQL: Introduce expression validation phase

### LINKED ISSUE #105425: ESQL: Introduce expression validation phase
### Description

Once constructed ESQL expressions have their typeResolved and if successful get executed. However due to the way a query works, this doesn't address the case where the argument is not properly replaced and thus the expression to be validated before actual execution:
`row a = 1 | eval b = a - 2 | eval y = date_trunch(@timestamp, b)`
B is an invalid argument to date_trunc since it's negative however until the plan gets optimized and the field properly replaced, that information is not available to date_trunc.
Furthermore postponing this information until execution is counter productive since invalid queries need to return errors while runtime/execution problem should return warning.

A solution to the above is to introduce a dedicated validate() phase that desired expressions can implement in order to verify the foldability and validity of their arguments.

### PR DESCRIPTION
**Pull request by costin**
Introduce dedicated phase through Validate interface that allows
 interested expressions to perform post logical optimization
 verification for things like literals/foldables which might be hidden
 through references after analysis.

Fix #105425

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/Validations.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import org.elasticsearch.xpack.ql.common.Failure;
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.Expression.TypeResolution;
+import org.elasticsearch.xpack.ql.expression.TypeResolutions;
+
+public final class Validations {
+
+    private Validations() {}
+
+    public static Failure isLiteral(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/Validations.java:19]: I know it's a short method, but please add java doc to this.  When I hover it in Intellij, I don't see the code, I only see the java doc.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalVerifier.java [hunk 2]
```diff
@@ -30,27 +24,25 @@ public final class LogicalVerifier {
     private LogicalVerifier() {}
 
     /** Verifies the optimized logical plan. */
-    public Collection<Failure> verify(LogicalPlan plan) {
-        Set<Failure> failures = new LinkedHashSet<>();
+    public Failures verify(LogicalPlan plan) {
+        Set<Failure> set = new LinkedHashSet<>();
+        Failures f = new Failures(set);
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalVerifier.java:29]: Nit: I think shortening the variable name to a single letter here decreases the readability.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/common/Failures.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+public class Failures {
+
+    private final Collection<Failure> failures;
+
+    public Failures() {
+        this(new ArrayList<>());
+    }
+
+    public Failures(Collection<Failure> failures) {
+        this.failures = failures;
```
Comment (not-napoleon) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/common/Failures.java:26]: Slightly dangerous - the `add` method assumes the collection is mutable, but we don't know that the collection passed in here is.  In practice, probably not a huge deal, but we should at least put some java doc on it to indicate the parameter is expected to be mutable.  If we want to be defensive, we could just make a mutable deep copy here too.

Or just don't include this constructor.

**Review by not-napoleon — APPROVED** [review]
LGTM

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/common/Failures.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ */
+
+package org.elasticsearch.xpack.ql.common;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Objects;
+
+/**
+ * Glorified list for managing {@link Failure}s.
```
Comment (alex-spies) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/common/Failures.java:15]: ++ to the nice description :D

**Review by luigidellaquila — APPROVED** [review]
LGTM

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/common/Failures.java (hunk 3 above)
Comment (alex-spies) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/common/Failures.java:26]: ++, `add` below might have surprisingly different behavior depending on whether we call the constructor with a list or a set of failures.

Maybe `Failures.failures` should be an `ArrayList<Failure>` for simplicity (or a set, if we want set semantics), and we just copy in this constructor?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/Validations.java (hunk 1 above)
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/Validations.java:19]: I think `isLiteral` may give the wrong impression.
```suggestion
    public static Failure isFoldable(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
```

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalVerifier.java [hunk 5]
```diff
@@ -30,27 +24,25 @@ public final class LogicalVerifier {
...
-                if (resolution.unresolved()) {
-                    failures.add(fail(e, resolution.message()));
-                }
-                resolution = isFoldable(e.to(), e.sourceText(), FOURTH);
-                if (resolution.unresolved()) {
-                    failures.add(fail(e, resolution.message()));
-                }
-            });
+            if (set.isEmpty()) {
+                // post optimization folding check
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalVerifier.java:37]: This comment is becoming out-of-date; I'd remove it.

**Review by alex-spies — APPROVED** [review]
LGTM, but maybe we want to ensure that `Failures.add` behaves consistently irrespectively of which `Collection<Failure>` we build the `Failures` object from.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/common/Failures.java (hunk 3 above)
Comment (costin) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/common/Failures.java:26]: I've encapsulated the code better as the list is not used elsewhere. 
As a side-note, internally the ESQL code should NOT be defensive or paranoid and rely on copying where necessary.

**Pull request merged**